src/storable.rs

```diff
@@ -492,6 +495,105 @@ where
     };
 }
 
+// Encodes `entry_bytes` size followed with `entry_bytes` into `bytes`
+// starting from the index `start`.
+// When encoding is saved in `bytes` on indices `[start, end)`
+// the function will return index `end` - the first index after
+// `start` that is not occupied with encoding.
+fn encode_with_size<T>(entry_bytes: &[u8], bytes: &mut [u8], start: usize) -> usize
+where
+    T: Storable,
+{
+    let size_len = get_num_bytes_required_to_store_size::<T>();
+    let actual_size = entry_bytes.len();
+
+    encode_size::<T>(&mut bytes[start..start + size_len], actual_size);
+
+    bytes[start + size_len..start + size_len + actual_size].copy_from_slice(entry_bytes);
+
+    start + actual_size + size_len
+}
+
+// Deserialize the struct starting at index `start` in `bytes`.
+// When serialized struct is saved in `bytes` on indices `[start, end)` the
+// function will return deserialized struct and index `end` - the first index
+// after `start` that is not occupied with the serialization of the struct.
+fn deserialize_with_size<T>(bytes: &[u8], start: usize) -> (T, usize)
+where
+    T: Storable,
+{
+    let size_len = get_num_bytes_required_to_store_size::<T>();
+    let actual_size = decode_size::<T>(&bytes[start..start + size_len]);
+
+    let a = T::from_bytes(Cow::Borrowed(
+        &bytes[start + size_len..start + size_len + actual_size],
+    ));
+    (a, start + actual_size + size_len)
+}
+
+impl<A, B, C> Storable for (A, B, C)
```

Comment:
1. TBH, I do not buy this argument. If a developer wants to change Bound from Bounded to Unbounded, that is a different type, so the developer should make it backward compatible. So it can load type as Bounded convert it to Unbounded and save it as Unbounded.
Even more, if we want to make it work in that case we will do even worse. There is no way we can support such behavior with the current serialization of the tuple (A, B) when A and B are bounded because we rely on the Bounds of A and B to determine size. So we will be left with inconsistent behavior because changing the bound will be allowed for (A, B, C) but not for (A, B).
2. I do not think such an operation will be useful on tuples that way, with that exact encoding, counter-example:
   A1 = "10", B1 = "1", A2 = "1", B2 = "11"
   so if we have tuples (A1, B1), (A2, B2)
   and we compare them (using >), expecting that we first compare the first element of the tuple followed by the second element first tuple will be bigger since A1 > A2, but (A1, B1).to_bytes() < (A2, B2).to_bytes() since the second bit in first tuple is "0" while in second is "1" and they have the same size.

Reply:
> TBH, I do not buy this argument. If a developer wants to change Bound from Bounded to Unbounded, that is a different type.

This is already a reality. The new btreemap was designed in such a way that allow developers to seamlessly upgrade and remove the bound on their types, and this is something we already support. If we now decide to not support this, we basically tell developers that they can't remove the bounds from their types and will have to create a new btreemap and migrate their data - it's too late to change that (nor is that best for a developer experience). I agree two-element tuples don't support it in its current form, but we can address that issue separately.

> I do not think such an operation will be useful on tuples that way, with that exact encoding, counter-example:
A1 = "10", B1 = "1", A2 = "1", B2 = "11"
so if we have tuples (A1, B1), (A2, B2)

That's a good counter-example. Perhaps there isn't a way to preserve that byte-sorting property in the encoding schema.

Reply:
This is implemented in [PR](https://github.com/dfinity/stable-structures/pull/202).